packages/vm/lib/evm/memory.ts

```diff
@@ -54,12 +54,18 @@ export default class Memory {
    * @param size - How many bytes to read
    */
   read(offset: number, size: number): Buffer {
-    const loaded = this._store.slice(offset, offset + size)
-    // Fill the remaining length with zeros
-    for (let i = loaded.length; i < size; i++) {
-      loaded[i] = 0
+    const returnBuffer = Buffer.allocUnsafe(size)
+    // Copy the stored "buffer" from memory into the return Buffer
+
+    const loaded = Buffer.from(this._store.slice(offset, offset + size))
+    returnBuffer.fill(loaded, 0, loaded.length)
+
+    if (loaded.length < size) {
+      // fill the remaining part of the Buffer with zeros
+      returnBuffer.fill(0, loaded.length, size)
     }
-    return Buffer.from(loaded)
+
+    return returnBuffer
```

Comment:
Just for my understanding, since I've read through both code versions a couple of times now and would assume that both would work: what has been changed/fixed here?

Reply:
There is a KECCAK test in the Homestead suite which requires one to hash ~1.5GB of data (`randomStateTest94`). The old approach uses too much memory (I assume because of how JS-arrays are implemented; one keeps pushing values into this array, thus expanding it). It runs out of memory on my pc (not sure if it also runs OOM on the CI) which has 16GB of RAM (also tested it with bumping the amount of RAM node can access). 

The issue is thus not that this changes any functionality; this however uses less memory as the new approach does not go OOM on my pc.

Reply:
Ah, nice explanation, thanks!

(and also cool "digging-deep")